File: app/src/main/java/com/eds/a2appstudiointerviewassignment/AddWebLinkDialog.java
package com.eds.a2appstudiointerviewassignment;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.webkit.URLUtil;
import android.widget.Button;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import com.google.android.material.textfield.TextInputEditText;

public class AddWebLinkDialog extends DialogFragment{

    public static String TAG = "AddWebLinkDialog";

    public interface AddWebLinkDialogListener {
        void clickPositiveButton(String url);
    }

    AddWebLinkDialogListener listener;

    public AddWebLinkDialog() {
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        try {
            listener = (AddWebLinkDialogListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(getActivity().toString() + " must implement AddWebLinkDialogListener");
        }
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        // Get the layout inflater
        LayoutInflater inflater = requireActivity().getLayoutInflater();
        View view = inflater.inflate(R.layout.dialog_add_weblink, null);
        TextInputEditText editText = view.findViewById(R.id.textInputEditText);
        Button buttonPositive = view.findViewById(R.id.buttonPositive);
        Button buttonNegative = view.findViewById(R.id.buttonNegative);
        editText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                String text = s.toString();
                if (text.length() > 5 && !URLUtil.isValidUrl(text)) {
                    editText.setError("Must be valid URL");
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
        buttonPositive.setOnClickListener(v -> {
            String userData = editText.getText().toString();
            if (URLUtil.isValidUrl(userData)) {
                listener.clickPositiveButton(userData);
                dismiss();
            }
        });
        buttonNegative.setOnClickListener(v -> {
            dismiss();
        });
        builder.setTitle("Add").setView(view);
        return builder.create();
    }
}

File: app/src/main/java/com/eds/a2appstudiointerviewassignment/dialogs/AddWebLinkDialog.kt
package com.eds.a2appstudiointerviewassignment.dialogs

import android.app.Dialog
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.webkit.URLUtil
import android.widget.Button
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import com.eds.a2appstudiointerviewassignment.R
import com.google.android.material.textfield.TextInputEditText

class AddWebLinkDialog(private val onPressedPositiveButton: (String) -> Unit): DialogFragment(){

    private lateinit var editText: TextInputEditText
    private lateinit var buttonPositive: Button
    private lateinit var buttonNegative: Button

    companion object {
        val TAG = "AddWebLinkDialog"
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val builder = AlertDialog.Builder(requireActivity())
        val view = requireActivity().layoutInflater.inflate(R.layout.dialog_add_weblink, null)

        editText = view.findViewById(R.id.textInputEditText)
        buttonPositive = view.findViewById(R.id.buttonPositive)
        buttonNegative = view.findViewById(R.id.buttonNegative)

        val onTextChanged = object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {}

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                val text = s.toString()
                if (!URLUtil.isValidUrl(text)) editText.error = "Must be valid URL"
            }
        }

        editText.addTextChangedListener(onTextChanged)

        buttonPositive.setOnClickListener {
            if (URLUtil.isValidUrl(editText.text.toString())) {
                onPressedPositiveButton(editText.text.toString())
                dismiss()
            }
        }

        buttonNegative.setOnClickListener {
            dismiss()
        }

        builder.setTitle("Add WebLink").setView(view)

        return builder.create()
    }
}